base/compiler/abstractinterpretation.jl

```diff
@@ -3063,45 +3285,116 @@ function override_effects(effects::Effects, override::EffectsOverride)
         nortcall = override.nortcall ? true : effects.nortcall)
 end
 
-isdefined_globalref(g::GlobalRef) = !iszero(ccall(:jl_globalref_boundp, Cint, (Any,), g))
-isdefinedconst_globalref(g::GlobalRef) = isconst(g) && isdefined_globalref(g)
+world_range(ir::IRCode) = ir.valid_worlds
+world_range(ci::CodeInfo) = WorldRange(ci.min_world, ci.max_world)
+world_range(compact::IncrementalCompact) = world_range(compact.ir)
 
-function abstract_eval_globalref_type(g::GlobalRef)
-    if isdefinedconst_globalref(g)
-        return Const(ccall(:jl_get_globalref_value, Any, (Any,), g))
+function force_binding_resolution!(g::GlobalRef)
+    # Force resolution of the binding
+    # TODO: This will go away once we switch over to fully partitioned semantics
+    ccall(:jl_globalref_boundp, Cint, (Any,), g)
+    return nothing
+end
+
+function abstract_eval_globalref_type(g::GlobalRef, src::Union{CodeInfo, IRCode, IncrementalCompact}, retry_after_resolve::Bool=true)
+    worlds = world_range(src)
+    partition = lookup_binding_partition(min_world(worlds), g)
+    partition.max_world < max_world(worlds) && return Any
+    while is_some_imported(binding_kind(partition))
+        imported_binding = partition_restriction(partition)::Core.Binding
+        partition = lookup_binding_partition(min_world(worlds), imported_binding)
+        partition.max_world < max_world(worlds) && return Any
+    end
+    if is_some_guard(binding_kind(partition))
+        if retry_after_resolve
+            # This method is surprisingly hot. For performance, don't ask the runtime to resolve
+            # the binding unless necessary - doing so triggers an additional lookup, which though
+            # not super expensive is hot enough to show up in benchmarks.
+            force_binding_resolution!(g)
+            return abstract_eval_globalref_type(g, src, false)
+        end
+        # return Union{}
```

Comment:
What's the TODO remaining to return `Union{}` here?

Reply:
Ah, I guess `# We do not currently assume an invalidation for guard -> defined transitions`?

Reply:
yes